src/components/Table/utils/RowState.ts

```diff
@@ -152,8 +150,8 @@ export class RowState {
   }
 
   private get visibleChildren(): RowState[] {
-    // The keptGroup should treat all of its children as visible, as this makes select/unselect all work.
-    if (this.row.kind === KEPT_GROUP) return this.children ?? [];
+    // The keptGroup is special and it's children are the dynamically kept rows
+    if (this.row.kind === KEPT_GROUP) return this.states.keptRows;
```

Comment:
Previously I'd been patching up `keptGroupRow.children` inside of every `setRows` call, which is why it was lucky/necessary for `setRows` to keep running all the time.

This makes the keptGroupRow.children reactive/calculated, instead of having to be imperatively updated, which is pretty cute imo. :-)